### DISCUSSION TITLE
Return type inline in return

### DISCUSSION THREAD

**Discussion by arthur-tacca** [sphinx-doc/sphinx#13125] [Feature requests and ideas]
I know I'm worrying over a triviality, but I find it a bit offensive that the return type has its own line, when the type of parameters and exceptions is inline:

![image](https://github.com/user-attachments/assets/c7274378-a4c9-4935-9fe2-07dfa37a25b7)

I'm happy to work around this by manually putting the return type in the description of `:return:` but then it looks very different to in the parameters and exceptions. (Actually even those are different from each other - bold vs italic - but this is monospace and boxed so looks even more different):

![image](https://github.com/user-attachments/assets/17981839-38dc-42e2-aba3-d681b5039864)

I think it would be good to be able to fold return type into the return item automatically. I'd be interested in any workaround ideas too.

**Comment by AA-Turner** [comment]
Are you using `:rtype:`? Please provide a reproducible example.

A

**Reply by arthur-tacca** [reply]
Yes the above example was just using `:rtype:` manually:

```rst
.. function:: my_test_fn(param1, param2)

   Example of documentation for a function.

   :param param1: Description of first parameter.
   :type param1: int | str
   :param param2: The second parameter.
   :type param2: ResultCapture
   :return: The result of the function.
   :rtype: ResultCapture
   :raise TaskNotDoneException: When an error happens.
```

But the same thing happens with autodoc and `autodoc_typehints = "description"`. The second example is using a normal reference in `:return:`:

```rst
   :return: :class:`ResultCapture` -- The result of the function.
```

Edit: Oh, and I should mention that I'm using the ReadTheDocs theme, though I don't think that fundamentally changes anything.

**Reply by AA-Turner** [reply]
Does using `:return ResultCapture: The result of the function.` work? 

xref https://www.sphinx-doc.org/en/master/usage/domains/python.html#info-field-lists

A

**Reply by arthur-tacca** [reply]
I can't test or post a screenshot right now, but I'm pretty sure I tried it earlier and it didn't work. I think that's because [`:return:` is defined](https://github.com/sphinx-doc/sphinx/blob/eb337fec0acd8ce978caa0663f60f3a6afd76322/sphinx/domains/python/_object.py#L176) to be a [`sphinx.util.docfields.Field`](https://github.com/sphinx-doc/sphinx/blob/eb337fec0acd8ce978caa0663f60f3a6afd76322/sphinx/util/docfields.py#L41) with `has_arg=False`.

**Reply by arthur-tacca** [reply]
Actually the temptation is to monkey patch that list with an extra entry that is a `PyGroupedField` like `:raise:`. I'll give that a go later. It wouldn't help if you're using autodoc though.

**Comment by arthur-tacca** [comment]
As a workaround, it can be fixed by switching the type of `:return:` to be the same sort of field as `:exception:`. This can be done by putting this in this in the `conf.py`:

```python
from sphinx.domains.python import PyObject, PyGroupedField

for i, f in enumerate(PyObject.doc_field_types):
    if f.name == "returnvalue":
        PyObject.doc_field_types[i] = PyGroupedField(
            f.name, label=f.label, names=f.names, rolename="class", can_collapse=True
        )
```

Then this Sphinx code:

```rst
.. function:: my_test_fn(param1, param2)

   Example of documentation for a function.

   :param param1: Description of first parameter.
   :type param1: int | str
   :param param2: The second parameter.
   :type param2: typing.Callable[[int | str, bool], ResultCapture]
   :return ResultCapture: The result of the function.
   :raise TaskNotDoneException: When an error happens.
```

gives this output:

![image](https://github.com/user-attachments/assets/c9f4a51d-453f-4af2-a730-b4ee3caaa497)

As a bonus (depending on your point of view), it will now automatically turn multiple entries into a list. For example, this input:

```rst
.. function:: my_test_fn(param1, param2)

   Example of documentation for a function.

   :param param1: Description of first parameter.
   :type param1: int | str
   :param param2: The second parameter.
   :type param2: ResultCapture
   :return ResultCapture: The result of the function.
   :return None: If nothing to return.
   :raise TaskNotDoneException: When an error happens.
```

gives this output:

![image](https://github.com/user-attachments/assets/b78d33a6-a2f9-4a59-bfca-89d544fbfa1a)

I originally thought of adding a totally new type of field, rather than patching the existing `:return:` meaning, but I've realised that it's harmless to do it this way: if you don't pass an argument then it's the same as it was before. It might even be worth updating Sphinx to do (although I get there could be a backwards-compatibility worry, even if it is presumably very rare to have multiple `:return:` fields in existing projects).

None of this solves the autodoc issue by itself. That would need additional changes. But this change would at least unlock a mechanism that an autodoc change could use (under a config option of course).

**Reply by arthur-tacca** [reply]
I've finally figured out a solution that will make this work for autodoc return type annotations too. Before the signature is processed, I delete the return annotation data so that autodoc doesn't add the `:rtype:` field. Then, during docstring processing, I patch it into the `:return:` field. 

```python
import re
from sphinx.util.typing import stringify_annotation

obj_returns = {}

def record_return_type(app, obj, bound_method):
    try:
        # Record the return annotation for this object, but also delete it so that autodoc does not
        # see it, which stops it from going into a redundant :rtype: directive.
        obj_returns[obj] = obj.__annotations__.pop("return")
    except AttributeError:
        pass
    except KeyError:
        pass

def put_in_return_type(app, what, name, obj, options, lines):
    # Bail out if this object never had a return type stored.
    if obj not in obj_returns:
        return
    # Format the string for this return type.
    rtype = stringify_annotation(obj_returns[obj], mode="smart").replace("\\", "\\\\")
    # For every line, if there is a bare :return: (no existing type) then add the return type.
    lines[:] = [
        re.sub(
            r"^(\s*):\s*returns?\s*:",
            rf"\1:return {rtype}:",
            line,
            re.IGNORECASE,
        )
        for line in lines
    ]


def setup(app):
    app.connect("autodoc-before-process-signature", record_return_type)
    app.connect("autodoc-process-docstring", put_in_return_type)
```

With this code, if your `:return:`field (or fields) already has an explicitly specified type then the return type annotation is ignored, which is probably what you want. 

This approach has the disadvantage that if you set [`autodoc_typehints="both"`](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#confval-autodoc_typehints), so that type hints still show in the function signature too, then the return type will be stripped from the signature as a side effect. The easiest fix for this is probably to get rid of the `autodoc-before-process-signature` handler, and instead prevent `:rtype:` creation by monkey patching `modify_field_list()` and `augment_descriptions_with_types()` in [`sphinx/ext/autodoc/typehints.py`](https://github.com/sphinx-doc/sphinx/blob/master/sphinx/ext/autodoc/typehints.py) to remove the `"return"` key from the `annotation` parameter. Then, when you patch the docstring, you can just read the annotation directly from the object. I didn't put that as the primary approach because it's more fragile too future changes in autodoc's internal design.

(It was surprisingly tough to come up with a solution. It didn't help that the autodoc signature events only apply to the types shown in the signature line, not the ones copied to the description. Also, fixing up rtype after it's been applied is almost impossible because autodoc puts type hints in the description by directly manipulating the docutils tree after it's been formed.)
